android/src/main/java/com/swmansion/gesturehandler/core/HoverGestureHandler.kt

```diff
@@ -0,0 +1,114 @@
+import android.view.MotionEvent
+import android.view.View
+import android.view.ViewGroup
+import com.swmansion.gesturehandler.react.RNViewConfigurationHelper
+
+class HoverGestureHandler : GestureHandler<HoverGestureHandler>() {
+  private var handler: Handler? = null
+  private var finishRunnable = Runnable { finish() }
+
+  private infix fun isAncestorOf(other: GestureHandler<*>): Boolean {
+    var current: View? = other.view
+
+    while (current != null) {
+      if (current == this.view) {
+        return true
+      }
+
+      current = current.parent as? View
+    }
+
+    return false
+  }
+
+  private fun isViewDisplayedOverAnother(view: View, other: View, rootView: View = view.rootView): Boolean? {
+    if (rootView == other) {
+      return true
+    } else if (rootView == view) {
+      return false
+    } else if (rootView is ViewGroup) {
+      for (i in 0 until rootView.childCount) {
+        val child = viewConfigHelper.getChildInDrawingOrderAtIndex(rootView, i)
+        return isViewDisplayedOverAnother(view, other, child) ?: continue
+      }
+    }
+
+    return null
+  }
+
+  override fun shouldBeCancelledBy(handler: GestureHandler<*>): Boolean {
+    if (handler is HoverGestureHandler) {
+      if (!(handler isAncestorOf this)) {
+        isViewDisplayedOverAnother(handler.view!!, this.view!!)?.let {
+          return it
+        }
+      }
+    }
+
+    return super.shouldBeCancelledBy(handler)
+  }
+
+  override fun shouldRequireToWaitForFailure(handler: GestureHandler<*>): Boolean {
+    if (handler is HoverGestureHandler) {
+      if (!(this isAncestorOf handler) && !(handler isAncestorOf this)) {
+        isViewDisplayedOverAnother(this.view!!, handler.view!!)?.let {
+          return it
+        }
+      }
+    }
+
+    return super.shouldRequireToWaitForFailure(handler)
+  }
+
+  override fun shouldRecognizeSimultaneously(handler: GestureHandler<*>): Boolean {
+    if (handler is HoverGestureHandler && (this isAncestorOf handler || handler isAncestorOf this)) {
+      return true
+    }
+
+    return super.shouldRecognizeSimultaneously(handler)
+  }
+
+  override fun onHandle(event: MotionEvent, sourceEvent: MotionEvent) {
+    if (event.action == MotionEvent.ACTION_DOWN) {
+      handler?.removeCallbacksAndMessages(null)
+      handler = null
+    } else if (event.action == MotionEvent.ACTION_UP) {
+      if (!isWithinBounds) {
+        finish()
+      }
+    }
+  }
+
+  override fun onHandleHover(event: MotionEvent, sourceEvent: MotionEvent) {
+    if (event.action == MotionEvent.ACTION_HOVER_EXIT) {
+      if (handler == null) {
+        handler = Handler(Looper.getMainLooper())
+      }
+
+      handler!!.postDelayed(finishRunnable, 4)
+    } else if (!isWithinBounds) {
+      finish()
+    } else if (event.action == MotionEvent.ACTION_HOVER_MOVE || event.action == MotionEvent.ACTION_HOVER_ENTER) {
+      if (this.state == STATE_UNDETERMINED) {
+        begin()
+        activate()
+      }
+    }
```

Comment:
Here I would also change it to avoid chaining if...else (and maybe combine last if statement with its inner if, but I'm not so sure about this one).

```suggestion
if (event.action == MotionEvent.ACTION_HOVER_EXIT) {
  if (handler == null) {
    handler = Handler(Looper.getMainLooper())
  }

  handler!!.postDelayed(finishRunnable, 4)
  return
}

if (!isWithinBounds) {
  finish()
  return
} 

if (event.action == MotionEvent.ACTION_HOVER_MOVE || event.action == MotionEvent.ACTION_HOVER_ENTER) {
  if (this.state == STATE_UNDETERMINED) {
      begin()
      activate()
    }
}
```

Reply:
I've tried to use [`when`](https://kotlinlang.org/docs/control-flow.html#when-expression) and in this case it doesn't look that bad (https://github.com/software-mansion/react-native-gesture-handler/pull/2455/commits/1a95954dbff666c337357771e9e989dfb47ee176)

Reply:
Yes, it looks much better now 🎉